ts/private/ts_project_worker.js

```diff
@@ -120,14 +173,60 @@ function createFilesystemTree(root, inputs) {
         }
         const dirs = [];
         for (const np in node) {
-            if (Object.keys(node[np]).length > 0) {
+            if (node[np][Type] == TYPE.DIR) {
                 dirs.push(np);
             }
         }
         return dirs
     }
 
-    return { add, remove, fileExists, directoryExists, readDirectory, getDirectories }
+    function notifyWatchers(parts, part, type, eventType) {
+        const dest_parts = [...parts, part];
+        if (type == TYPE.FILE) {
+            notifyWatcher(parts, dest_parts, eventType);
+        } else if (type == TYPE.DIR) {
+            notifyWatcher(parts, parts, eventType);
+        }
+        notifyWatcher(dest_parts, dest_parts, eventType);
+    }
+
+    function notifyWatcher(parent, parts, eventType) {
+        let node = getWatcherNode(parent, watchingTree);
+        if (typeof node == "object" && Watcher in node) {
+            for (const callback of node[Watcher]) {
+                callback(parts.join(path.sep), eventType);
+            }
+        }
+    }
+
+    function getWatcherNode(parts) {
+        let node = watchingTree;
+        for (const part of parts) {
+            if (!(part in node)) {
+                return undefined;
+            }
+            node = node[part];
+        }
+        return node;
+    }
+
+    function watch(p, callback) {
+        const parts = p.split(path.sep);
+        let node = watchingTree;
+        for (const part of parts) {
+            if (!(part in node)) {
+                node[part] = {};
+            }
+            node = node[part];
+        }
+        if (!(Watcher in node)) {
+            node[Watcher] = new Array();
+        }
+        let i = node[Watcher].push(callback) - 1;
+        return () => node[Watcher].splice(i, 1);
```

Comment:
Can't the index change? Unless the nodes are added/removed in a fifo order I think the indexes will change, and in that case just use push/pop instead of indexes.

I assume this has to be `node[Watcher] = node[Watcher].filter(...)`?

Reply:
Good catch. you are absolutely right.

Reply:
Can you think of an easy test to add that reproduces it?

Reply:
I think so

Reply:
Added a new test case and fixed!